File: simulation/mnist.py
#!/Users/alessio/anaconda3/bin/python3

import numpy as np

dictDecoder = {
    
    "8": "ubyte",	# 无符号字节
    "9": "byte",	# 有符号字节
    "11": ">i2",	# 大端序16位整数
    "12": ">i4",	# 大端序32位整数
    "13": ">f4",	# 大端序32位浮点数
    "14": ">f8"		# 大端序64位浮点数（双精度）
}

def loadDataset(images, labels):

    ''' 
    将 MNIST 数据集中的图像和标签加载到两个 NumPy 数组中

    输入参数：

        1) images: 字符串，对应包含 28x28 黑白图像的 MNIST 文件名，
        这些图像以 idx 格式存储

        2) labels: 字符串，对应包含标签的 MNIST 文件名，
        标签以无符号字节序列的形式存储

    返回值：

        1) imgArray: 二维 NumPy 数组，包含从 MNIST 文件中读取的所有图像
        每个图像存储为形状为 (1,784) 的 NumPy 数组，便于用作神经网络的输入，
        其中每个输入神经元对应一个像素

        2) labelsArray: NumPy 数组，包含从 MNIST 文件中读取的所有标签，
        以整数形式存储
    '''

    # 将图像文件的全部内容加载到内存缓冲区
    imgBuffer = readFile(images)

    # 创建用于训练/测试的图像数组
    imgArray = idxBufferToArray(imgBuffer)

    # 将标签文件的全部内容加载到内存缓冲区
    labelsBuffer = readFile(labels)

    # 创建用于训练/测试的标签数组
    labelsArray = idxBufferToArray(labelsBuffer)

    return imgArray, labelsArray


def readFile(filename):

    '''
    读取二进制文件的全部内容并将其存储在内存缓冲区中

    输入参数：
        filename: 字符串，对应要读取的文件名

    返回值：
        readData: 缓冲区，其中存储了文件的全部内容，以字节序列形式表示
    '''
    with open(filename, "r+b") as f:
        readData = f.read()

    return readData


def idxBufferToArray(buffer):

    '''
    将以 idx 格式存储数据的二进制缓冲区转换为 NumPy 数组

    输入参数：
        buffer: 以 idx 格式编码数据的字节序列
        可以通过调用 readFile() 函数从 idx 文件中读取数据获得

    返回值：
        data: NumPy 数组，仅包含从 idx 缓冲区中提取的数据

    magic number和维度信息会被读取并用于确定输出数组的形状，但它们不会出现在输出数据数组中
    '''

    # 读取并解码magic number
    dtype, dataDim = magicNumber(buffer)

    # magic number占4字节，每个维度占4字节
    offset = 4*dataDim+4

    # 读取并存储数据结构的所有维度
    dimensions = readDimensions(buffer, dataDim)

    # 将数据存储到适当形状的 NumPy 数组中
    data = loadData(buffer, dtype, offset, dimensions)

    return data


def magicNumber(buffer):

    '''
    读取并解码magic number

    输入参数：
        buffer: 以 idx 格式编码数据的字节序列
        可以通过调用 readFile() 函数从 idx 文件中读取数据获得

    返回值：
        1) dtype: 表示数据类型的字符串。详情见 dictDecoder
        2) dataDim: 存储在 idx 缓冲区中数据结构的维度数量

    注意，decodeDataType 期望输入为十进制数，而数据类型在magic number中以十六进制值编码
    幸运的是，NumPy 的 frombuffer 函数会直接将十六进制值转换为对应的十进制值
    '''

    # 将magic number读取为四个单独的字节
    mn = np.frombuffer(buffer, dtype="ubyte", count=4)

    # 解码数据类型字节
    dtype = decodeDataType(mn[2])

    # 解码数据维度字节
    dataDim = mn[3]

    return dtype, dataDim


def decodeDataType(intCode):

    '''
    解码magic number中编码数据类型的字节。

    输入参数：
        intCode: 从magic number的第二低有效字节读取的整数值，以十进制格式表示

    返回值：

        函数返回一个字符串，对应magic number中编码的数据类型。详情见 dictDecoder
    '''
    return dictDecoder[str(intCode)]


def readDimensions(buffer, dataDim):

    '''
    从 idx 缓冲区中读取适当数量的维度。维度以32位整数存储
    输入参数：
        1) buffer: 以 idx 格式编码数据的字节序列。
        可以通过调用 readFile() 函数从 idx 文件中读取数据获得
        2) dataDim: 要读取的维度数量。可以通过调用 magicNumber() 函数获得

    返回值：
        函数返回一个形状为 (1, dataDim) 的 NumPy 数组，包含所有维度
    '''
    return np.frombuffer(buffer, dtype=">u4", count=dataDim, offset=4)


def loadData(buffer, dtype, offset, dimensions):

    '''
    从以 idx 格式存储数据的缓冲区中读取完整的数据结构

    输入参数：
        1) buffer: 以 idx 格式编码数据的字节序列
        可以通过调用 readFile() 函数从 idx 文件中读取数据获得
        2) dtype: 编码要读取数据类型的字符串
        3) offset: 表示数据在 idx 缓冲区中起始位置的偏移量，以字节为单位
        4) dimensions: 包含要读取数据结构所有维度的 NumPy 数组

    返回值：
        函数返回一个具有适当形状的 NumPy 数组，包含所有数据
    '''

    # 将数据存储为单一维度的 NumPy 数组
    data = np.frombuffer(buffer, dtype=dtype, count=np.prod(dimensions),
        offset=offset) 

    # 将数据重塑为多维 NumPy 数组并返回
    return reshapeData(data, dimensions)



def reshapeData(data, dimensions):

    '''
    重塑 NumPy 数组

    新形状通过解释从 idx 缓冲区读取的维度获得

    如果缓冲区只包含单一维度，则意味着数据以单一数组形式存储，
    输出 NumPy 数组也将具有单一维度

    如果维度是多个，则输出数组将被重塑为二维 NumPy 数组，
    将高阶维度展平为单一维度

    输入参数：
        1) data: 包含所有数据的 NumPy 数组
        2) dimensions: 包含所有维度的 NumPy 数组

    输出值：
        data: 重塑后的 NumPy 数组
    '''
    
    # 仅在必要时重塑
    if dimensions.size > 1:

        # 重塑为二维数组
        arrayDim = np.prod(dimensions[1:])
        data = np.reshape(data, (dimensions[0], arrayDim))

    return data

File: src/simulation/mnist.rs
use byteorder::{BigEndian, ReadBytesExt};
use ndarray::{Array2, ArrayD};
use std::fs::File;
use std::io::{Cursor, Read};

// 数据类型解码器
fn decode_data_type(int_code: u32) -> String {
    match int_code {
        8 => "u8".to_string(),   // 无符号字节
        9 => "i8".to_string(),   // 有符号字节
        11 => "i16".to_string(), // 大端 16 位整数
        12 => "i32".to_string(), // 大端 32 位整数
        13 => "f32".to_string(), // 大端 32 位浮点数
        14 => "f64".to_string(), // 大端 64 位浮点数
        _ => panic!("未知的数据类型代码: {}", int_code),
    }
}

// 读取文件内容
fn read_file(filename: &str) -> Vec<u8> {
    let mut file = File::open(filename).expect("无法打开文件");
    let mut buffer = Vec::new();
    file.read_to_end(&mut buffer).expect("无法读取文件");
    buffer
}

// 读取和解码魔术数
fn magic_number(buffer: &[u8]) -> (String, u8) {
    let mut cursor = Cursor::new(buffer);
    let mn = cursor.read_u32::<BigEndian>().expect("无法读取魔术数");
    let data_type_code = (mn >> 8) & 0xFF; // 提取数据类型字节
    let data_dim = (mn & 0xFF) as u8; // 提取维度数字节
    let dtype = decode_data_type(data_type_code);
    (dtype, data_dim)
}

// 读取维度
fn read_dimensions(buffer: &[u8], data_dim: u8) -> Vec<u32> {
    let mut cursor = Cursor::new(&buffer[4..]);
    let mut dimensions = Vec::new();
    for _ in 0..data_dim {
        let dim = cursor.read_u32::<BigEndian>().expect("无法读取维度");
        dimensions.push(dim);
    }
    dimensions
}

// 加载数据
fn load_data(buffer: &[u8], dtype: &str, offset: usize, dimensions: &[u32]) -> ArrayD<f64> {
    let total_elements = dimensions.iter().product::<u32>() as usize;
    let data_slice = &buffer[offset..];

    let data: Vec<f64> = match dtype {
        "u8" => data_slice.iter().map(|&x| x as f64).collect(),
        "i8" => data_slice.iter().map(|&x| x as i8 as f64).collect(),
        "i16" => {
            let mut cursor = Cursor::new(data_slice);
            (0..total_elements)
                .map(|_| cursor.read_i16::<BigEndian>().unwrap() as f64)
                .collect()
        }
        "i32" => {
            let mut cursor = Cursor::new(data_slice);
            (0..total_elements)
                .map(|_| cursor.read_i32::<BigEndian>().unwrap() as f64)
                .collect()
        }
        "f32" => {
            let mut cursor = Cursor::new(data_slice);
            (0..total_elements)
                .map(|_| cursor.read_f32::<BigEndian>().unwrap() as f64)
                .collect()
        }
        "f64" => {
            let mut cursor = Cursor::new(data_slice);
            (0..total_elements)
                .map(|_| cursor.read_f64::<BigEndian>().unwrap())
                .collect()
        }
        _ => panic!("不支持的数据类型: {}", dtype),
    };

    let shape: Vec<usize> = dimensions.iter().map(|&x| x as usize).collect();
    ArrayD::from_shape_vec(shape, data).expect("无法创建数组")
}

// 重塑数据
fn reshape_data(data: ArrayD<f64>, dimensions: &[u32]) -> ArrayD<f64> {
    if dimensions.len() > 1 {
        let array_dim = dimensions[1..].iter().product::<u32>() as usize;
        let shape = (dimensions[0] as usize, array_dim);
        data.into_shape(shape)
            .expect("重塑数组失败")
            .to_owned()
            .into_dyn()
    } else {
        data
    }
}

// 将 IDX 缓冲区转换为数组
fn idx_buffer_to_array(buffer: &[u8]) -> ArrayD<f64> {
    let (dtype, data_dim) = magic_number(buffer);
    let offset = 4 * data_dim as usize + 4; // 魔术数和维度后的数据偏移量
    let dimensions = read_dimensions(buffer, data_dim);
    let data = load_data(buffer, &dtype, offset, &dimensions);
    reshape_data(data, &dimensions)
}

// 加载数据集
pub fn load_dataset(images: &str, labels: &str) -> (Array2<f64>, ArrayD<f64>) {
    let img_buffer = read_file(images);
    let img_array = idx_buffer_to_array(&img_buffer);

    let labels_buffer = read_file(labels);
    let labels_array = idx_buffer_to_array(&labels_buffer);

    (
        img_array.into_dimensionality().expect("图像数组维度错误"),
        labels_array,
    )
}

#[cfg(test)]
mod tests {
    use super::*;
    use image::{GrayImage, Luma};
    use ndarray::Axis;

    // 将 Array2<f64> 转换为 GrayImage
    fn array_to_image(array: &Array2<f64>) -> GrayImage {
        let (height, width) = array.dim();
        let mut img = GrayImage::new(width as u32, height as u32);
        for y in 0..height {
            for x in 0..width {
                let pixel = array[[y, x]] as u8; // 将 f64 转换为 u8
                img.put_pixel(x as u32, y as u32, Luma([pixel]));
            }
        }
        img
    }

    // 展示 MNIST 图像
    fn show_mnist_image(image: &Array2<f64>) {
        let img = array_to_image(image);
        img.save("mnist_image.png").expect("无法保存图像");
        println!("图像已保存为 mnist_image.png");
    }

    #[test]
    fn test_load_mnist_image() {
        let images = "src/simulation/mnist/t10k-images-idx3-ubyte";
        let labels = "src/simulation/mnist/t10k-labels-idx1-ubyte";
        let (img_array, _labels_array) = load_dataset(images, labels);

        // 获取第一张图像，形状为 [784] 的一维数组
        let image_flat = img_array.index_axis(Axis(0), 0).to_owned();

        // 将一维数组重塑为 28x28 的二维数组
        let image = image_flat
            .into_shape((28, 28))
            .expect("无法将图像重塑为 28x28");

        // 调用函数，传递二维数组的引用
        show_mnist_image(&image);
    }
}
